cipher_suite.go

```diff
@@ -84,62 +90,80 @@ func CipherSuiteName(id CipherSuiteID) string {
-		return newCipherSuiteTLSEcdheEcdsaWithAes128Ccm()
+		return NewCipherSuiteTLSEcdheEcdsaWithAes128Ccm()
 	case TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8:
-		return newCipherSuiteTLSEcdheEcdsaWithAes128Ccm8()
+		return NewCipherSuiteTLSEcdheEcdsaWithAes128Ccm8()
 	case TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256:
-		return &cipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{}
+		return &CipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{}
 	case TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256:
-		return &cipherSuiteTLSEcdheRsaWithAes128GcmSha256{}
+		return &CipherSuiteTLSEcdheRsaWithAes128GcmSha256{}
 	case TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA:
-		return &cipherSuiteTLSEcdheEcdsaWithAes256CbcSha{}
+		return &CipherSuiteTLSEcdheEcdsaWithAes256CbcSha{}
 	case TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA:
-		return &cipherSuiteTLSEcdheRsaWithAes256CbcSha{}
+		return &CipherSuiteTLSEcdheRsaWithAes256CbcSha{}
 	case TLS_PSK_WITH_AES_128_CCM:
-		return newCipherSuiteTLSPskWithAes128Ccm()
+		return NewCipherSuiteTLSPskWithAes128Ccm()
 	case TLS_PSK_WITH_AES_128_CCM_8:
-		return newCipherSuiteTLSPskWithAes128Ccm8()
+		return NewCipherSuiteTLSPskWithAes128Ccm8()
 	case TLS_PSK_WITH_AES_128_GCM_SHA256:
-		return &cipherSuiteTLSPskWithAes128GcmSha256{}
+		return &CipherSuiteTLSPskWithAes128GcmSha256{}
+	case TLS_ECDH_ANON_WITH_AES_128_CBC_SHA256:
+		return NewCipherSuiteTLSEcdhAnonWithAes128CbcSha256(TLS_ECDH_ANON_WITH_AES_128_CBC_SHA256)
 	}
 	return nil
 }
 
 // CipherSuites we support in order of preference
-func defaultCipherSuites() []cipherSuite {
-	return []cipherSuite{
-		&cipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
-		&cipherSuiteTLSEcdheRsaWithAes128GcmSha256{},
-		&cipherSuiteTLSEcdheEcdsaWithAes256CbcSha{},
-		&cipherSuiteTLSEcdheRsaWithAes256CbcSha{},
+func defaultCipherSuites() []CipherSuite {
+	return []CipherSuite{
+		&CipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
+		&CipherSuiteTLSEcdheRsaWithAes128GcmSha256{},
+		&CipherSuiteTLSEcdheEcdsaWithAes256CbcSha{},
+		&CipherSuiteTLSEcdheRsaWithAes256CbcSha{},
 	}
 }
 
-func allCipherSuites() []cipherSuite {
-	return []cipherSuite{
-		newCipherSuiteTLSEcdheEcdsaWithAes128Ccm(),
-		newCipherSuiteTLSEcdheEcdsaWithAes128Ccm8(),
-		&cipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
-		&cipherSuiteTLSEcdheRsaWithAes128GcmSha256{},
-		&cipherSuiteTLSEcdheEcdsaWithAes256CbcSha{},
-		&cipherSuiteTLSEcdheRsaWithAes256CbcSha{},
-		newCipherSuiteTLSPskWithAes128Ccm(),
-		newCipherSuiteTLSPskWithAes128Ccm8(),
-		&cipherSuiteTLSPskWithAes128GcmSha256{},
+func allCipherSuites() []CipherSuite {
+	return []CipherSuite{
+		NewCipherSuiteTLSEcdheEcdsaWithAes128Ccm(),
+		NewCipherSuiteTLSEcdheEcdsaWithAes128Ccm8(),
+		&CipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
+		&CipherSuiteTLSEcdheRsaWithAes128GcmSha256{},
+		&CipherSuiteTLSEcdheEcdsaWithAes256CbcSha{},
+		&CipherSuiteTLSEcdheRsaWithAes256CbcSha{},
+		NewCipherSuiteTLSPskWithAes128Ccm(),
+		NewCipherSuiteTLSPskWithAes128Ccm8(),
+		&CipherSuiteTLSPskWithAes128GcmSha256{},
```

Comment:
Why do we need all cipher suites to be public? This means we are adding all these structs and methods to the public api of pion/dtls.

Reply:
As @Sean-Der mentioned we want to move them to the separate package. And i want to allow customize cipher like: https://github.com/plgd-dev/sdk/blob/justWorks/local/core/otm/just-works/client.go#L45 because OCF has own cipher number for ANON.

Reply:
> As @Sean-Der mentioned we want to move them to the separate package

:+1: Ok to that. Since it will be a public API it should be cleaned up a little bit. For example creation of CipherSuite `&CipherSuiteXXX{}` vs `NewCipherSuiteXXX`.

>And i want to allow customize cipher like: https://github.com/plgd-dev/sdk/blob/justWorks/local/core/otm/just-works/client.go#L45 because OCF has own cipher number for ANON.

Another option will be to use `CipherSuitesFactory` to append a new custom `CipherSuite` that replicates the code from `NewCipherSuiteTLSEcdhAnonWithAes128CbcSha256` but with a different cipher number. This approach means having code redundacy for a specific use case, but the pion/dtls API will be more simple.